src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java

```diff
@@ -52,4 +93,278 @@ private long countGermplasm(UUID programId, String trialDbId) throws ApiExceptio
+        List<Map<String, Object>> exportRows = new ArrayList<>(rowByOUId.values());
+        if (fileType.equals(FileType.CSV)){
+            downloadFile = CSVWriter.writeToDownload(columns, exportRows, fileType);
+        } else {
+            downloadFile = ExcelWriter.writeToDownload("Dataset Export", columns, exportRows, fileType);
+        }
+
+        String envFilenameFragment = params.getEnvironments() == null ? "All Environments" : params.getEnvironments();
+        String fileName = makeFileName(experiment, program, envFilenameFragment) + fileType.getExtension();
+        return new DownloadFile(fileName, downloadFile);
+    }
+
+    private void addBrAPIObsToRecords(
+            List<BrAPIObservation> dataset,
+            BrAPITrial experiment,
+            Program program,
+            List<BrAPIObservationUnit> ous,
+            Map<String, BrAPIStudy> studyByDbId,
+            Map<String, Map<String, Object>> rowByOUId,
+            boolean includeTimestamp,
+            List<BrAPIObservationVariable> obsVars) throws ApiException, DoesNotExistException {
+        for (BrAPIObservation obs: dataset) {
+            BrAPIObservationUnit ou = ous.stream()
+                    .filter(unit -> obs.getObservationUnitDbId().equals(unit.getObservationUnitDbId()))
+                    .findAny()
+                    .orElseThrow(RuntimeException::new);
+            String ouId = getOUId(ou);
+
+            // if there is a row with that ouId then just add the obs var data and timestamp to the row
+            if (rowByOUId.get(ouId) != null) {
+                addObsVarDataToRow(rowByOUId.get(ouId), obs, includeTimestamp, obsVars, program);
+            } else {
+
+                // otherwise make a new row
+                Map<String, Object> row = createExportRow(experiment, program, ou, studyByDbId);
+                addObsVarDataToRow(row, obs, includeTimestamp, obsVars, program);
+                rowByOUId.put(ouId, row);
+            }
+        }
+    }
+
+    private String getOUId(BrAPIObservationUnit ou) {
+        BrAPIExternalReference ouXref = Utilities.getExternalReference(
+                        ou.getExternalReferences(),
+                        String.format("%s/%s", referenceSource, ExternalReferenceSource.OBSERVATION_UNITS.getName()))
+                .orElseThrow(() -> new RuntimeException("observation unit id not found"));
+        return ouXref.getReferenceID();
+    }
+
+    private void addObsVarDataToRow(
+            Map<String, Object> row,
+            BrAPIObservation obs,
+            boolean includeTimestamp,
+            List<BrAPIObservationVariable> obsVars,
+            Program program) {
+        // get observation variable for BrAPI observation
+        BrAPIObservationVariable var = obsVars.stream()
+                .filter(obsVar -> obs.getObservationVariableName().equals(obsVar.getObservationVariableName()))
+                .collect(Collectors.toList()).get(0);
```

Comment:
It would also be worth changing this to make a map of `variableByDbId`, and have that map be created and held in `addBrAPIObsToRecords`.  This will also reduce the runtime complexity.  This method could then be updated to accept a single `BrAPIObservationVariable` instead of a list of `BrAPIObservationVariables`

Reply:
done

Reply:
done